src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,761 @@
+ *
+ *     public static class Bar {
+ *         public Bar(Foo foo) {
+ *              // ...
+ *         }
+ *     }
+ *
+ *     private static final Supplier<Foo> FOO = StableValue.supplier(Foo::new);
+ *     private static final Supplier<Bar> BAR = StableValue.supplier(() -> new Bar(FOO.get()));
+ *
+ *     public static Foo foo() {
+ *         return FOO.get();
+ *     }
+ *
+ *     public static Bar bar() {
+ *         return BAR.get();
+ *     }
+ *
+ * }
+ *}
+ * Calling {@code bar()} will create the {@code Bar} singleton if it is not already
+ * created. Upon such a creation, the dependent {@code Foo} will first be created if
+ * the {@code Foo} does not already exist.
+ * <p>
+ * Here is another example where a more complex dependency graph is created in which
+ * integers in the Fibonacci delta series are lazily computed:
+ * {@snippet lang = java:
+ * public final class Fibonacci {
+ *
+ *     private Fibonacci() {}
+ *
+ *     private static final int MAX_SIZE_INT = 46;
+ *
+ *     private static final IntFunction<Integer> FIB =
+ *             StableValue.intFunction(MAX_SIZE_INT, Fibonacci::fib);
+ *
+ *     public static int fib(int n) {
+ *         return n < 2
+ *                 ? n
+ *                 : FIB.apply(n - 1) + FIB.apply(n - 2);
+ *     }
+ *
+ * }
+ *}
+ * Both {@code FIB} and {@code Fibonacci::fib} recurse into each other. Because the
+ * stable int function {@code FIB} caches intermediate results, the initial
+ * computational complexity is reduced from exponential to linear compared to a
+ * traditional non-caching recursive fibonacci method. Once computed, the VM is free to
+ * constant-fold expressions like {@code Fibonacci.fib(5)}.
+ * <p>
+ * The fibonacci example above is a dependency graph with no circular dependencies (i.e.,
+ * it is a dependency tree):
+ *{@snippet lang=text :
+ *
+ *              ___________fib(5)____________
+ *             /                             \
+ *       ____fib(4)____                  ____fib(3)____
+ *      /              \                /              \
+ *    fib(3)          fib(2)          fib(2)          fib(1)
+ *   /     \         /     \         /     \
+ * fib(2) fib(1)   fib(1) fib(0)   fib(1) fib(0)
+ *}
+ *
+ * If there are circular dependencies in a dependency graph, a stable value will
+ * eventually throw an {@linkplain IllegalStateException} upon referencing elements in
+ * a circularity.
+ *
+ * <h2 id="thread-safety">Thread Safety</h2>
+ * The content of a stable value is guaranteed to be set at most once. If competing
+ * threads are racing to set a stable value, only one update succeeds, while other updates
+ * are blocked until the stable value becomes set.
+ * <p>
+ * The at-most-once write operation on a stable value that succeeds
+ * (e.g. {@linkplain #trySet(Object) trySet()})
+ * {@linkplain java.util.concurrent##MemoryVisibility <em>happens-before</em>}
+ * any subsequent read operation (e.g. {@linkplain #orElseThrow()}) that is successful.
```

Comment:
Is `isSet()` also considered to be a "read operation"? I would assume so, but that requires a bit of reading between the lines. Maybe `isSet()` could be added as another parenthetical example after `orElseThrow()` here to squash that particular ambiguity.

Reply:
Very good observation. I will clarify this. Thanks!